File: examples/healthcare-portal-extraction/rest/python/healthcare_portal_extraction.py
# Extracts patient records from a sandbox healthcare portal using BQL.
#
# Install: pip install requests
# Run:     python healthcare_portal_extraction.py

import requests

query = """
mutation HealthcarePortal {
  goto(url: "https://scraping-sandbox.netlify.app/clarity-health", waitUntil: networkIdle) {
    status
  }
  waitForSelector(selector: ".patient-record", timeout: 10000) {
    time
  }
  patients: mapSelector(selector: ".patient-record") {
    name: mapSelector(selector: ".patient-name") { innerText }
    dob: mapSelector(selector: ".patient-dob") { innerText }
    provider: mapSelector(selector: ".patient-provider") { innerText }
    nextAppt: mapSelector(selector: ".patient-appointment") { innerText }
    status: mapSelector(selector: ".patient-status") { innerText }
  }
}
"""

response = requests.post(
    'https://production-sfo.browserless.io/chromium/bql',
    params={'token': 'YOUR_API_TOKEN_HERE'},
    json={'query': query, 'variables': {}, 'operationName': 'HealthcarePortal'},
)

data = response.json()['data']
for patient in data['patients']:
    name = patient['name'][0]['innerText'] if patient['name'] else ''
    dob = patient['dob'][0]['innerText'] if patient['dob'] else ''
    provider = patient['provider'][0]['innerText'] if patient['provider'] else ''
    status = patient['status'][0]['innerText'] if patient['status'] else ''
    print(f'{name} | {dob} | {provider} | {status}')

File: examples/healthcare-portal-extraction/rest/java/HealthcarePortalExtraction.java
// Extracts patient records from a sandbox healthcare portal using BQL.
//
// Run: javac HealthcarePortalExtraction.java && java HealthcarePortalExtraction

import java.net.URI;
import java.net.http.*;

public class HealthcarePortalExtraction {
    public static void main(String[] args) throws Exception {
        String token = "YOUR_API_TOKEN_HERE";
        String endpoint = "https://production-sfo.browserless.io/chromium/bql?token=" + token;

        String query = "mutation HealthcarePortal { goto(url: \\"https://scraping-sandbox.netlify.app/clarity-health\\", waitUntil: networkIdle) { status } waitForSelector(selector: \\".patient-record\\", timeout: 10000) { time } patients: mapSelector(selector: \\".patient-record\\") { name: mapSelector(selector: \\".patient-name\\") { innerText } dob: mapSelector(selector: \\".patient-dob\\") { innerText } provider: mapSelector(selector: \\".patient-provider\\") { innerText } nextAppt: mapSelector(selector: \\".patient-appointment\\") { innerText } status: mapSelector(selector: \\".patient-status\\") { innerText } } }";

        String payload = "{\"query\": \"" + query + "\", \"variables\": {}, \"operationName\": \"HealthcarePortal\"}";

        HttpClient client = HttpClient.newHttpClient();
        HttpRequest request = HttpRequest.newBuilder()
            .uri(URI.create(endpoint))
            .header("Content-Type", "application/json")
            .POST(HttpRequest.BodyPublishers.ofString(payload))
            .build();

        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        System.out.println(response.body());
    }
}
